File: Core/Source/NSAttributedString+DTCoreText.m
//
//  NSAttributedString+DTCoreText.m
//  DTCoreText
//
//  Created by Oliver Drobnik on 2/1/12.
//  Copyright (c) 2012 Drobnik.com. All rights reserved.
//

#import "DTCompatibility.h"
#import "NSAttributedString+DTCoreText.h"
#import "DTHTMLWriter.h"
#import "DTCoreTextConstants.h"
#import "DTCoreTextFontDescriptor.h"
#import "DTCoreTextParagraphStyle.h"
#import "DTCSSListStyle.h"
#import "DTImageTextAttachment.h"
#import "NSString+Paragraphs.h"
#import "NSDictionary+DTCoreText.h"
#import "NSAttributedStringRunDelegates.h"

#import <DTFoundation/NSURL+DTComparing.h>

#if TARGET_OS_IPHONE
#import "UIFont+DTCoreText.h"
#endif

@implementation NSAttributedString (DTCoreText)

#pragma mark Text Attachments
- (NSArray *)textAttachmentsWithPredicate:(NSPredicate *)predicate class:(Class)class
{
	if (![self length])
	{
		return nil;
	}
	
	NSMutableArray *foundAttachments = [NSMutableArray array];
	
	NSRange entireRange = NSMakeRange(0, [self length]);
	[self enumerateAttribute:NSAttachmentAttributeName inRange:entireRange options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired usingBlock:^(DTTextAttachment *attachment, NSRange range, BOOL *stop) {
		
		if (attachment == nil)
		{
			// no attachment value
			return;
		}
		
		if (predicate && ![predicate evaluateWithObject:attachment])
		{
			// doesn't fit predicate, next
			return;
		}
		
		if (class && ![attachment isKindOfClass:class])
		{
			// doesn't fit class, next
			return;
		}
		
		[foundAttachments addObject:attachment];
	}];
	
	if ([foundAttachments count])
	{
		return foundAttachments;
	}
	
	return nil;
}


#pragma mark Calculating Ranges

- (NSInteger)itemNumberInTextList:(DTCSSListStyle *)list atIndex:(NSUInteger)location
{
	NSRange effectiveRange;
	NSArray *textListsAtIndex = [self attribute:DTTextListsAttribute atIndex:location effectiveRange:&effectiveRange];
	
	if (!textListsAtIndex)
	{
		return 0;
	}
	
	// get outermost list
	DTCSSListStyle *outermostList = [textListsAtIndex objectAtIndex:0];
	
	// get the range of all lists
	NSRange totalRange = [self rangeOfTextList:outermostList atIndex:location];
	
	// get naked NSString
    NSString *string = [[self string] substringWithRange:totalRange];
	
    // entire string
    NSRange range = NSMakeRange(0, [string length]);
	
	NSMutableDictionary *countersPerList = [NSMutableDictionary dictionary];
	
	// enumerating through the paragraphs in the plain text string
    [string enumerateSubstringsInRange:range options:NSStringEnumerationByParagraphs usingBlock:^(NSString *substring, NSRange substringRange, NSRange enclosingRange, BOOL *stop)
     {
		 NSRange paragraphListRange;
		 NSArray *textLists = [self attribute:DTTextListsAttribute atIndex:substringRange.location + totalRange.location effectiveRange:&paragraphListRange];
		 
		 DTCSSListStyle *currentEffectiveList = [textLists lastObject];
		 
		 NSNumber *key = [NSNumber numberWithInteger:(NSInteger)currentEffectiveList]; // list address is identifier
		 NSNumber *currentCounterNum = [countersPerList objectForKey:key];
		 
		 NSInteger currentCounter=0;
		 
		 if (!currentCounterNum)
		 {
			 currentCounter = currentEffectiveList.startingItemNumber;
		 }
		 else
		 {
			 currentCounter = [currentCounterNum integerValue]+1;
		 }
		 
		 currentCounterNum = [NSNumber numberWithInteger:currentCounter];
		 [countersPerList setObject:currentCounterNum forKey:key];
		 
		 // calculate the actual range
		 NSRange actualRange = enclosingRange;  // includes a potential \n
		 actualRange.location += totalRange.location;
		 
		 if (NSLocationInRange(location, actualRange))
		 {
			 *stop = YES;
		 }
     }
     ];
	
	NSNumber *key = [NSNumber numberWithInteger:(NSInteger)list]; // list address is identifier
	NSNumber *currentCounterNum = [countersPerList objectForKey:key];
	
	return [currentCounterNum integerValue];
}

- (NSRange)_rangeOfObject:(id)object inArrayBehindAttribute:(NSString *)attribute atIndex:(NSUInteger)location
{
	@synchronized(self)
	{
		NSUInteger stringLength = [self length];
		NSUInteger searchIndex = location;
		
		NSArray *arrayAtIndex;
		
		NSRange totalRange = NSMakeRange(NSNotFound, 0);
		
		BOOL foundList = NO;
		
		do
		{
			NSRange effectiveRange;
			arrayAtIndex = [self attribute:attribute atIndex:searchIndex effectiveRange:&effectiveRange];
			
			if (!arrayAtIndex || [arrayAtIndex indexOfObjectIdenticalTo:object] == NSNotFound)
			{
				break;
			}
			
			searchIndex = effectiveRange.location;
			foundList = YES;
			
			// enhance found range
			if (totalRange.location == NSNotFound)
			{
				totalRange = effectiveRange;
			}
			else
			{
				totalRange = NSUnionRange(totalRange, effectiveRange);
			}
			
			if (searchIndex == 0)
			{
				// reached beginning of string
				break;
			}
			
			searchIndex--;
		}
		while (foundList);
		
		// if we didn't find the list at all, return
		if (!foundList)
		{
			return NSMakeRange(NSNotFound, 0);
		}
		
		// now search forward
		
		searchIndex = NSMaxRange(totalRange);
		
		while (searchIndex < stringLength)
		{
			NSRange effectiveRange;
			arrayAtIndex = [self attribute:attribute atIndex:searchIndex effectiveRange:&effectiveRange];
			
			if (!arrayAtIndex || [arrayAtIndex indexOfObjectIdenticalTo:object] == NSNotFound)
			{
				break;
			}
			
			searchIndex = NSMaxRange(effectiveRange);
			
			// enhance found range
			totalRange = NSUnionRange(totalRange, effectiveRange);
		}
		
		return totalRange;
	}
}

- (NSRange)rangeOfTextList:(DTCSSListStyle *)list atIndex:(NSUInteger)location
{
	NSParameterAssert(list);
	
	NSRange listRange = [self _rangeOfObject:list inArrayBehindAttribute:DTTextListsAttribute atIndex:location];
	
	if (listRange.location == NSNotFound)
	{
		// list was not found
		return listRange;
	}
	
	// extend list range to full paragraphs to be safe
	listRange = [self.string rangeOfParagraphsContainingRange:listRange parBegIndex:NULL parEndIndex:NULL];
	
	return listRange;
}

- (NSRange)rangeOfTextBlock:(DTTextBlock *)textBlock atIndex:(NSUInteger)location
{
	NSParameterAssert(textBlock);
	
	return [self _rangeOfObject:textBlock inArrayBehindAttribute:DTTextBlocksAttribute atIndex:location];
}

- (NSRange)rangeOfAnchorNamed:(NSString *)anchorName
{
	__block NSRange foundRange = NSMakeRange(NSNotFound, 0);
	
	[self enumerateAttribute:DTAnchorAttribute inRange:NSMakeRange(0, [self length]) options:0 usingBlock:^(NSString *value, NSRange range, BOOL *stop) {
		if ([value isEqualToString:anchorName])
		{
			*stop = YES;
			foundRange = range;
		}
	}];
	
	return foundRange;
}

- (NSRange)rangeOfLinkAtIndex:(NSUInteger)location URL:(NSURL * __autoreleasing*)URL
{
	NSRange rangeSoFar;
	
	NSURL *foundURL = [self attribute:DTLinkAttribute atIndex:location effectiveRange:&rangeSoFar];
	
	if (!foundURL)
	{
		return NSMakeRange(NSNotFound, 0);
	}
	
	// search towards beginning
	while (rangeSoFar.location>0)
	{
		NSRange extendedRange;
		NSURL *extendedURL = [self attribute:DTLinkAttribute atIndex:rangeSoFar.location-1 effectiveRange:&extendedRange];
		
		// abort search if key not found or value not identical
		if (!extendedURL || ![extendedURL isEqualToURL:foundURL])
		{
			break;
		}
		
		rangeSoFar = NSUnionRange(rangeSoFar, extendedRange);
	}
	
	NSUInteger length = [self length];
	
	// search towards end
	while (NSMaxRange(rangeSoFar)<length)
	{
		NSRange extendedRange;
		NSURL *extendedURL = [self attribute:DTLinkAttribute atIndex:NSMaxRange(rangeSoFar) effectiveRange:&extendedRange];
		
		// abort search if key not found or value not identical
		if (!extendedURL || ![extendedURL isEqualToURL:foundURL])
		{
			break;
		}
		
		rangeSoFar = NSUnionRange(rangeSoFar, extendedRange);
	}
	
	if (URL)
	{
		*URL = foundURL;
	}
	
	return rangeSoFar;
}

- (NSRange)rangeOfFieldAtIndex:(NSUInteger)location
{
    if (location<[self length])
    {
        // get range of prefix
        NSRange fieldRange;
        NSString *fieldAttribute = [self attribute:DTFieldAttribute atIndex:location effectiveRange:&fieldRange];
        
        if (fieldAttribute)
        {
            return fieldRange;
        }
    }
    
    return NSMakeRange(NSNotFound, 0);
}

#pragma mark HTML Encoding

#ifndef COVERAGE
// exclude method from coverage testing, those are just convenience methods

- (NSString *)htmlString
{
	// create a writer
	DTHTMLWriter *writer = [[DTHTMLWriter alloc] initWithAttributedString:self];
	
	// return it's output
	return [writer HTMLString];
}

- (NSString *)htmlFragment
{
	// create a writer
	DTHTMLWriter *writer = [[DTHTMLWriter alloc] initWithAttributedString:self];
	
	// return it's output
	return [writer HTMLFragment];
}

- (NSString *)plainTextString
{
	NSString *tmpString = [self string];
	
	return [tmpString stringByReplacingOccurrencesOfString:UNICODE_OBJECT_PLACEHOLDER withString:@""];
}

#endif

#pragma mark Generating Special Attributed Strings
+ (NSAttributedString *)prefixForListItemWithCounter:(NSUInteger)listCounter listStyle:(DTCSSListStyle *)listStyle listIndent:(CGFloat)listIndent attributes:(NSDictionary *)attributes
{
	// get existing values from attributes
	CTParagraphStyleRef paraStyle = (__bridge CTParagraphStyleRef)[attributes objectForKey:(id)kCTParagraphStyleAttributeName];
	CTFontRef font = (__bridge CTFontRef)[attributes objectForKey:(id)kCTFontAttributeName];
	
	DTCoreTextFontDescriptor *fontDescriptor = nil;
	DTCoreTextParagraphStyle *paragraphStyle = nil;
	
	if (paraStyle)
	{
		paragraphStyle = [DTCoreTextParagraphStyle paragraphStyleWithCTParagraphStyle:paraStyle];
		
		paragraphStyle.tabStops = nil;
		
		paragraphStyle.headIndent = listIndent;
		
		if (listStyle.type != DTCSSListStyleTypeNone)
		{
			// first tab is to right-align bullet, numbering against
			CGFloat tabOffset = paragraphStyle.headIndent - (CGFloat)5.0; // TODO: change with font size
			[paragraphStyle addTabStopAtPosition:tabOffset alignment:kCTRightTextAlignment];
		}
		
		// second tab is for the beginning of first line after bullet
		[paragraphStyle addTabStopAtPosition:paragraphStyle.headIndent alignment:kCTLeftTextAlignment];
	}
	
	if (font)
	{
		fontDescriptor = [DTCoreTextFontDescriptor fontDescriptorForCTFont:font];
	}
	
	NSMutableDictionary *newAttributes = [NSMutableDictionary dictionary];
	
	if (fontDescriptor)
	{
		// make a font without italic or bold
		DTCoreTextFontDescriptor *fontDesc = [fontDescriptor copy];
		
		fontDesc.boldTrait = NO;
		fontDesc.italicTrait = NO;
		
		font = [fontDesc newMatchingFont];
		
		if (font)
		{
#if DTCORETEXT_SUPPORT_NS_ATTRIBUTES && __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_5_1
			if (___useiOS6Attributes)
			{
				UIFont *uiFont = [UIFont fontWithCTFont:font];
				[newAttributes setObject:uiFont forKey:NSFontAttributeName];
				
				CFRelease(font);
			}
			else
#endif
			{
				[newAttributes setObject:CFBridgingRelease(font) forKey:(id)kCTFontAttributeName];
			}
		}
	}
	
	CGColorRef textColor = (__bridge CGColorRef)[attributes objectForKey:(id)kCTForegroundColorAttributeName];
	
	if (textColor)
	{
		[newAttributes setObject:(__bridge id)textColor forKey:(id)kCTForegroundColorAttributeName];
	}
#if DTCORETEXT_SUPPORT_NS_ATTRIBUTES
	else if (___useiOS6Attributes)
	{
		DTColor *uiColor = [attributes foregroundColor];
		
		if (uiColor)
		{
			[newAttributes setObject:uiColor forKey:NSForegroundColorAttributeName];
		}
	}
#endif
	
	// add paragraph style (this has the tabs)
	if (paragraphStyle)
	{
#if DTCORETEXT_SUPPORT_NS_ATTRIBUTES
		if (___useiOS6Attributes)
		{
			NSParagraphStyle *style = [paragraphStyle NSParagraphStyle];
			[newAttributes setObject:style forKey:NSParagraphStyleAttributeName];
		}
		else
#endif
		{
			CTParagraphStyleRef newParagraphStyle = [paragraphStyle createCTParagraphStyle];
			[newAttributes setObject:CFBridgingRelease(newParagraphStyle) forKey:(id)kCTParagraphStyleAttributeName];
		}
	}
	
	// add textBlock if there's one (this has padding and background color)
	NSArray *textBlocks = [attributes objectForKey:DTTextBlocksAttribute];
	if (textBlocks)
	{
		[newAttributes setObject:textBlocks forKey:DTTextBlocksAttribute];
	}
	
	// transfer all lists so that
	NSArray *lists = [attributes objectForKey:DTTextListsAttribute];
	if (lists)
	{
		[newAttributes setObject:lists forKey:DTTextListsAttribute];
	}
	
	// add a marker so that we know that this is a field/prefix
	[newAttributes setObject:DTListPrefixField forKey:DTFieldAttribute];
	
	NSString *prefix = [listStyle prefixWithCounter:listCounter];
	
	if (prefix)
	{
		DTImage *image = nil;
		
		if (listStyle.imageName)
		{
			image = [DTImage imageNamed:listStyle.imageName];
			
			if (!image)
			{
				// image invalid
				listStyle.imageName = nil;
				
				prefix = [listStyle prefixWithCounter:listCounter];
			}
		}
		
		NSMutableAttributedString *tmpStr = [[NSMutableAttributedString alloc] initWithString:prefix attributes:newAttributes];
		
		if (image)
		{
			// make an attachment for the image
			DTImageTextAttachment *attachment = [[DTImageTextAttachment alloc] init];
			attachment.image = image;
			attachment.displaySize = image.size;
			
#if DTCORETEXT_SUPPORT_NS_ATTRIBUTES && TARGET_OS_IPHONE
			// need run delegate for sizing
			CTRunDelegateRef embeddedObjectRunDelegate = createEmbeddedObjectRunDelegate(attachment);
			[newAttributes setObject:CFBridgingRelease(embeddedObjectRunDelegate) forKey:(id)kCTRunDelegateAttributeName];
#endif
			
			// add attachment
			[newAttributes setObject:attachment forKey:NSAttachmentAttributeName];
			
			if (listStyle.position == DTCSSListStylePositionInside)
			{
				[tmpStr setAttributes:newAttributes range:NSMakeRange(2, 1)];
			}
			else
			{
				[tmpStr setAttributes:newAttributes range:NSMakeRange(1, 1)];
			}
		}
		
		return tmpStr;
	}
	
	return nil;
}

@end

File: Sources/DTCoreTextSwift/NSAttributedString+DTCoreText.swift
//
//  NSAttributedString+DTCoreText.swift
//  DTCoreText
//
//  Created by Oliver Drobnik on 2/1/12.
//  Copyright (c) 2012 Drobnik.com. All rights reserved.
//
//  Migrated to Swift, April 2026.
//

import Foundation
import CoreText

#if os(iOS) || os(tvOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

// NS-style attributes are always used on iOS 16+ / macOS 13+.

// MARK: - NSAttributedString (DTCoreText)

extension NSAttributedString {

	// MARK: Text Attachments

	/// Retrieves the `TextAttachment` objects that match the given predicate.
	///
	/// With this method you can for example find all images that have a certain URL.
	///
	/// - Parameters:
	///   - predicate: The predicate to apply for filtering or `nil` to not filter by attachment.
	///   - theClass: The class that attachments need to have, or `nil` for all attachments regardless of class.
	/// - Returns: The filtered array of attachments, or `nil` if none found.
	@objc
	public func textAttachments(with predicate: NSPredicate?, `class` theClass: AnyClass?) -> [TextAttachment]? {
		guard self.length > 0 else {
			return nil
		}

		var foundAttachments = [TextAttachment]()

		let entireRange = NSRange(location: 0, length: self.length)
		self.enumerateAttribute(.attachment, in: entireRange, options: .longestEffectiveRangeNotRequired) { value, _, _ in
			guard let attachment = value as? TextAttachment else {
				return
			}

			if let predicate, !predicate.evaluate(with: attachment) {
				return
			}

			if let theClass, !attachment.isKind(of: theClass) {
				return
			}

			foundAttachments.append(attachment)
		}

		return foundAttachments.isEmpty ? nil : foundAttachments
	}

	// MARK: Calculating Ranges

	/// Returns the index of the item at the given location within the list.
	///
	/// - Parameters:
	///   - list: The text list.
	///   - location: The location of the item.
	/// - Returns: The index within the list.
	@objc
	public func itemNumber(in list: CSSListStyle, at location: Int) -> Int {
		var effectiveRange = NSRange(location: 0, length: 0)
		guard let textListsAtIndex = self.attribute(NSAttributedString.Key(rawValue: DTTextListsAttribute), at: location, effectiveRange: &effectiveRange) as? [CSSListStyle] else {
			return 0
		}

		// get outermost list
		let outermostList = textListsAtIndex[0]

		// get the range of all lists
		let totalRange = self.rangeOfTextList(outermostList, at: location)

		// get naked NSString
		let string = (self.string as NSString).substring(with: totalRange)

		// entire string
		let range = NSRange(location: 0, length: (string as NSString).length)

		var countersPerList = [NSNumber: NSNumber]()

		// enumerating through the paragraphs in the plain text string
		(string as NSString).enumerateSubstrings(in: range, options: .byParagraphs) { _, substringRange, enclosingRange, stop in
			var paragraphListRange = NSRange(location: 0, length: 0)
			let textLists = self.attribute(NSAttributedString.Key(rawValue: DTTextListsAttribute), at: substringRange.location + totalRange.location, effectiveRange: &paragraphListRange) as? [CSSListStyle]

			guard let currentEffectiveList = textLists?.last else {
				return
			}

			// list address is identifier
			let key = NSNumber(integerLiteral: Int(bitPattern: Unmanaged.passUnretained(currentEffectiveList).toOpaque()))
			let currentCounterNum = countersPerList[key]

			var currentCounter: Int

			if currentCounterNum == nil {
				currentCounter = currentEffectiveList.startingItemNumber()
			} else {
				currentCounter = currentCounterNum!.intValue + 1
			}

			countersPerList[key] = NSNumber(value: currentCounter)

			// calculate the actual range
			var actualRange = enclosingRange // includes a potential \n
			actualRange.location += totalRange.location

			if NSLocationInRange(location, actualRange) {
				stop.pointee = true
			}
		}

		// list address is identifier
		let key = NSNumber(integerLiteral: Int(bitPattern: Unmanaged.passUnretained(list).toOpaque()))
		let currentCounterNum = countersPerList[key]

		return currentCounterNum?.intValue ?? 0
	}

	/// Private helper to find the range of an object in an array-valued attribute.
	private func _rangeOfObject(_ object: AnyObject, inArrayBehindAttribute attribute: String, at location: Int) -> NSRange {
		let lock = NSLock()
		lock.lock()
		defer { lock.unlock() }

		let stringLength = self.length
		var searchIndex = location

		var totalRange = NSRange(location: NSNotFound, length: 0)

		var foundList = false

		repeat {
			var effectiveRange = NSRange(location: 0, length: 0)
			let arrayAtIndex = self.attribute(NSAttributedString.Key(rawValue: attribute), at: searchIndex, effectiveRange: &effectiveRange) as? [AnyObject]

			if arrayAtIndex == nil || arrayAtIndex!.firstIndex(where: { $0 === object }) == nil {
				break
			}

			searchIndex = effectiveRange.location
			foundList = true

			// enhance found range
			if totalRange.location == NSNotFound {
				totalRange = effectiveRange
			} else {
				totalRange = NSUnionRange(totalRange, effectiveRange)
			}

			if searchIndex == 0 {
				// reached beginning of string
				break
			}

			searchIndex -= 1
		} while foundList

		// if we didn't find the list at all, return
		if !foundList {
			return NSRange(location: NSNotFound, length: 0)
		}

		// now search forward
		searchIndex = NSMaxRange(totalRange)

		while searchIndex < stringLength {
			var effectiveRange = NSRange(location: 0, length: 0)
			let arrayAtIndex = self.attribute(NSAttributedString.Key(rawValue: attribute), at: searchIndex, effectiveRange: &effectiveRange) as? [AnyObject]

			if arrayAtIndex == nil || arrayAtIndex!.firstIndex(where: { $0 === object }) == nil {
				break
			}

			searchIndex = NSMaxRange(effectiveRange)

			// enhance found range
			totalRange = NSUnionRange(totalRange, effectiveRange)
		}

		return totalRange
	}

	/// Returns the range of the given text list that contains the given location.
	///
	/// - Parameters:
	///   - list: The text list.
	///   - location: The location in the text.
	/// - Returns: The range of the given text list containing the location.
	@objc
	public func rangeOfTextList(_ list: CSSListStyle, at location: Int) -> NSRange {
		precondition(list !== nil as AnyObject?, "list must not be nil")

		var listRange = _rangeOfObject(list, inArrayBehindAttribute: DTTextListsAttribute, at: location)

		if listRange.location == NSNotFound {
			// list was not found
			return listRange
		}

		// extend list range to full paragraphs to be safe
		listRange = (self.string as NSString).range(ofParagraphsContaining: listRange, parBeg: nil, parEnd: nil)

		return listRange
	}

	/// Returns the range of the given text block that contains the given location.
	///
	/// - Parameters:
	///   - textBlock: The text block.
	///   - location: The location in the text.
	/// - Returns: The range of the given text block containing the location.
	@objc
	public func rangeOfTextBlock(_ textBlock: TextBlock, at location: Int) -> NSRange {
		precondition(textBlock !== nil as AnyObject?, "textBlock must not be nil")

		return _rangeOfObject(textBlock, inArrayBehindAttribute: DTTextBlocksAttribute, at: location)
	}

	/// Returns the range of the given href anchor.
	///
	/// - Parameters:
	///   - anchorName: The name of the anchor.
	/// - Returns: The range of the given anchor.
	@objc
	public func rangeOfAnchorNamed(_ anchorName: String) -> NSRange {
		var foundRange = NSRange(location: NSNotFound, length: 0)

		self.enumerateAttribute(NSAttributedString.Key(rawValue: DTAnchorAttribute), in: NSRange(location: 0, length: self.length), options: []) { value, range, stop in
			if let value = value as? String, value == anchorName {
				stop.pointee = true
				foundRange = range
			}
		}

		return foundRange
	}

	/// Returns the range of the hyperlink at the given index.
	///
	/// - Parameters:
	///   - location: The location to query.
	///   - URL: On output, the URL that is found at this location.
	/// - Returns: The range of the given hyperlink.
	@objc
	public func rangeOfLink(at location: Int, url: AutoreleasingUnsafeMutablePointer<NSURL?>?) -> NSRange {
		var rangeSoFar = NSRange(location: 0, length: 0)

		guard let foundURL = self.attribute(NSAttributedString.Key(rawValue: DTLinkAttribute), at: location, effectiveRange: &rangeSoFar) as? NSURL else {
			return NSRange(location: NSNotFound, length: 0)
		}

		// search towards beginning
		while rangeSoFar.location > 0 {
			var extendedRange = NSRange(location: 0, length: 0)
			let extendedURL = self.attribute(NSAttributedString.Key(rawValue: DTLinkAttribute), at: rangeSoFar.location - 1, effectiveRange: &extendedRange) as? NSURL

			// abort search if key not found or value not identical
			if extendedURL == nil || !extendedURL!.isEqual(foundURL) {
				break
			}

			rangeSoFar = NSUnionRange(rangeSoFar, extendedRange)
		}

		let length = self.length

		// search towards end
		while NSMaxRange(rangeSoFar) < length {
			var extendedRange = NSRange(location: 0, length: 0)
			let extendedURL = self.attribute(NSAttributedString.Key(rawValue: DTLinkAttribute), at: NSMaxRange(rangeSoFar), effectiveRange: &extendedRange) as? NSURL

			// abort search if key not found or value not identical
			if extendedURL == nil || !extendedURL!.isEqual(foundURL) {
				break
			}

			rangeSoFar = NSUnionRange(rangeSoFar, extendedRange)
		}

		url?.pointee = foundURL

		return rangeSoFar
	}

	/// Returns the range of a field at the given index.
	///
	/// - Parameter location: The location of the field.
	/// - Returns: The range of the field. If there is no field at this location it returns `{NSNotFound, 0}`.
	@objc
	public func rangeOfField(at location: Int) -> NSRange {
		if location < self.length {
			// get range of prefix
			var fieldRange = NSRange(location: 0, length: 0)
			let fieldAttribute = self.attribute(NSAttributedString.Key(rawValue: DTFieldAttribute), at: location, effectiveRange: &fieldRange) as? String

			if fieldAttribute != nil {
				return fieldRange
			}
		}

		return NSRange(location: NSNotFound, length: 0)
	}

	// MARK: HTML Encoding

	/// Converts the receiver into plain text.
	///
	/// This is different from the `string` method of `NSAttributedString` by also erasing
	/// placeholders for text attachments.
	///
	/// - Returns: The receiver converted to plain text.
	@objc
	public var plainTextString: String {
		return self.string.replacingOccurrences(of: UNICODE_OBJECT_PLACEHOLDER, with: "")
	}

	// MARK: Generating Special Attributed Strings

	/// Create a prefix for a paragraph in a list.
	///
	/// - Parameters:
	///   - listCounter: The value for the list item.
	///   - listStyle: The list style.
	///   - listIndent: The amount in px to indent the list.
	///   - attributes: The attribute dictionary for the text to be prefixed.
	/// - Returns: An attributed string with the list prefix.
	@objc
	public static func prefixForListItem(withCounter listCounter: UInt, listStyle: CSSListStyle, listIndent: CGFloat, attributes: [String: Any]) -> NSAttributedString? {
		// get existing values from attributes
		let ctParagraphStyleKey = NSAttributedString.Key(rawValue: kCTParagraphStyleAttributeName as String)
		let ctFontKey = NSAttributedString.Key(rawValue: kCTFontAttributeName as String)

		let paraStyle = attributes[kCTParagraphStyleAttributeName as String] as CFTypeRef?
		let fontRef = attributes[kCTFontAttributeName as String] as CFTypeRef?

		var fontDescriptor: CoreTextFontDescriptor?
		var paragraphStyle: CoreTextParagraphStyle?

		if let paraStyle {
			let ctParaStyle = unsafeBitCast(paraStyle, to: CTParagraphStyle.self)
			paragraphStyle = CoreTextParagraphStyle(ctParagraphStyle: ctParaStyle)

			paragraphStyle!.tabStops = nil
			paragraphStyle!.headIndent = listIndent

			if listStyle.type != .none {
				// first tab is to right-align bullet, numbering against
				let tabOffset = paragraphStyle!.headIndent - 5.0
				paragraphStyle!.addTabStop(atPosition: tabOffset, alignment: CTTextAlignment.right)
			}

			// second tab is for the beginning of first line after bullet
			paragraphStyle!.addTabStop(atPosition: paragraphStyle!.headIndent, alignment: CTTextAlignment.left)
		}

		if let fontRef {
			let ctFont = unsafeBitCast(fontRef, to: CTFont.self)
			fontDescriptor = CoreTextFontDescriptor(ctFont: ctFont)
		}

		var newAttributes = [NSAttributedString.Key: Any]()

		if let fontDescriptor {
			// make a font without italic or bold
			let fontDesc = fontDescriptor.copy() as! CoreTextFontDescriptor

			fontDesc.boldTrait = false
			fontDesc.italicTrait = false

			let newFont = fontDesc.newMatchingFont()

			if let newFont {
				if true {
					#if os(iOS) || os(tvOS)
					let uiFont = UIFont.font(withCTFont: newFont)
					newAttributes[.font] = uiFont
					#endif
				} else {
					newAttributes[ctFontKey] = newFont
				}
			}
		}

		let ctForegroundColorKey = NSAttributedString.Key(rawValue: kCTForegroundColorAttributeName as String)
		let textColor = attributes[kCTForegroundColorAttributeName as String] as CFTypeRef?

		if let textColor {
			newAttributes[ctForegroundColorKey] = textColor
		} else if true {
			let attributesDict = attributes as NSDictionary
			let uiColor = attributesDict.foregroundColor()

			if let uiColor {
				newAttributes[.foregroundColor] = uiColor
			}
		}

		// add paragraph style (this has the tabs)
		if let paragraphStyle {
			if true {
				let style = paragraphStyle.nsParagraphStyle()
				newAttributes[.paragraphStyle] = style
			} else {
				let newParagraphStyle = paragraphStyle.createCTParagraphStyle()
				newAttributes[ctParagraphStyleKey] = newParagraphStyle
			}
		}

		// add textBlock if there's one (this has padding and background color)
		if let textBlocks = attributes[DTTextBlocksAttribute] {
			newAttributes[NSAttributedString.Key(rawValue: DTTextBlocksAttribute)] = textBlocks
		}

		// transfer all lists
		if let lists = attributes[DTTextListsAttribute] {
			newAttributes[NSAttributedString.Key(rawValue: DTTextListsAttribute)] = lists
		}

		// add a marker so that we know that this is a field/prefix
		newAttributes[NSAttributedString.Key(rawValue: DTFieldAttribute)] = DTListPrefixField

		let prefix = listStyle.prefix(withCounter: Int(listCounter))

		guard let prefix else {
			return nil
		}

		let tmpStr = NSMutableAttributedString(string: prefix, attributes: newAttributes)

		#if os(iOS) || os(tvOS)
		var image: DTImage?

		if let imageName = listStyle.imageName {
			image = DTImage(named: imageName)

			if image == nil {
				// image invalid
				listStyle.imageName = nil
				// prefix was already obtained above; re-obtain is not needed since
				// the prefix string doesn't change from image name removal in this path
			}
		}

		if let image {
			// make an attachment for the image
			let attachment = ImageTextAttachment()
			attachment.image = image
			attachment.displaySize = image.size

			// need run delegate for sizing
			if let embeddedObjectRunDelegate = createEmbeddedObjectRunDelegate(attachment) {
				newAttributes[NSAttributedString.Key(rawValue: kCTRunDelegateAttributeName as String)] = embeddedObjectRunDelegate
			}

			// add attachment
			newAttributes[.attachment] = attachment

			if listStyle.position == .inside {
				tmpStr.setAttributes(newAttributes, range: NSRange(location: 2, length: 1))
			} else {
				tmpStr.setAttributes(newAttributes, range: NSRange(location: 1, length: 1))
			}
		}
		#endif

		return tmpStr
	}
}
